[FIX] web_editor: stop media-dialog from loading all db attachments

sdegueldre opened the pull request:
Before this commit, the media-dialog was reading all attachments that
matched certain criteria for each tab. On large databases, this results
in the search_read call timing out, preventing the media-dialog from
opening and making all images and documents un-editable.

This commit fixes that by reading the attachments page by page, and
only reading the next page as needed.

task-2198527

qsm-odoo commented on addons/web_editor/static/src/js/widgets/widgets.js:
```diff
@@ -136,6 +136,9 @@ var ImageDialog = Widget.extend({
         this.page = 0;
+        this.records = [];
+        this.needle = '';
+        this.per_page = this.IMAGES_PER_ROW * this.IMAGES_ROWS;
```
You can camelCase everything, it will be forward-ported.

qsm-odoo commented on addons/web_editor/static/src/js/widgets/widgets.js:
```diff
@@ -159,6 +162,12 @@ var ImageDialog = Widget.extend({
             self.page += $target.hasClass('previous') ? -1 : 1;
+            if ((self.page + 1) * self.per_page === self.records.length) {
+                // Preload next page if needed.
+                return self.proxy('fetch_page')(self.page + 1).then(function () {
```
Why the need of using proxy each time you call fetch_page?

sdegueldre replied:
Actually no need. I thought it was used because there was no auto-binding in 11.0 but it turns out it's just a specialized call-time bind for when you want to pass it as-is into a `then`, which I don't even do anymore.

qsm-odoo commented on addons/web_editor/static/src/js/widgets/widgets.js:
```diff
@@ -159,6 +162,12 @@ var ImageDialog = Widget.extend({
                 return;
             }
             self.page += $target.hasClass('previous') ? -1 : 1;
+            if ((self.page + 1) * self.per_page === self.records.length) {
```
Not sure to understand the condition. If you already have the record of the next page then you fetch them again?

sdegueldre replied:
If the last record on the current page is the last record overall, then I fetch the next page. `this.page` is 0 indexed, hence the `+ 1`. I think I'll just add a field to keep track of when I've read all matching records, it'll be much easier to understand that way.

sdegueldre commented:
@qsm-odoo updated according to what we discussed :)

sdegueldre commented:
>     * Current 11.0: 4 search_read on media dialog start (don't know why that many already)

There are 2 searches: one for images and one for documents, that much is normal. But there was a call to set_image in start that would do another unnecessary search, I got rid of that one so we're down to 2.

>     * Your branch: 8 search_read... wild guess is that we perform the current search_read 4 times by mistake and since it becomes 2 search_read per initial search with your implementation... you get 8 search_read.
> 
> 
> Could you investigate? Also would not it be easy to fetch the two pages in one _rpc by just adding an extra parameter to fetch_page, allowing to search for a page range ?
> 
> Also: not sure but reading your code I though that when reaching the before-last page, you already fetch the last page... but when going to the last page, I get a RPC again (while there is no next page?).

Fixed all of that. I was "pre-loading" the next page, but actually not using that at all, so I changed some things around. Now I just load one page + 1 records, that last record allows the rest of the code to behave like there is a next page, and when you go to the next page, we fetch the rest of it then. This allows me to keep most of the code untouched and I don't even need to keep track of a lastLoadedPage, which is nice.

qsm-odoo commented:
@sdegueldre Seems nice :) Could you just add a comment explaining why you fetch +1 record per page so that an innocent dev does not remove that +1 ? Thanks :)

sdegueldre commented:
@qsm-odoo done :+1:

sdegueldre commented:
> Sorry, again a change needed: add "Closes XXX" with XXX being the url of the issue this commit solves.

Should I use the short github form (`odoo/odoo#43113`) which github will linkify, or a full url? Also isn't it weird to say "closes: odoo/odoo#43113" when the issue is actually in a later version and shouldn't be closed until the fix is properly forward ported?

qsm-odoo commented:
Meh, I think you overthink this. That's just to mention the original issue, the fact that it could close the issue is not important we will forward-port anyway. If you want you can mention the issue with "See XXX" then close the issue by yourself later.

Personally, I always put the full URL so that I can access those from git log, but do as you want ;)

sdegueldre commented:
@qsm-odoo done :)

robodoo commented:
Staging failed: timed out (>120 minutes)

xmo-odoo commented:
@robodoo retry
